File: sdk-typescript/src/services/storage-provider-watch.ts
import * as chokidar from 'chokidar';
import * as fs from 'fs';
import * as path from 'path';

export interface FileChangeEvent {
  /** Workspace-relative path */
  path: string;
  /** Event type matching proto FileChangeType */
  eventType: string;
}

const DEFAULT_IGNORE_DIRS = new Set([
  '.git',
  'node_modules',
  '__pycache__',
  'target',
  'build',
  '.elevo',
]);

/** Coalescing window before flush (ms) */
const COALESCE_DELAY_MS = 50;
/** Max latency cap before forced flush (ms) */
const MAX_LATENCY_MS = 200;
/** Degraded mode poll interval (ms) */
const DEGRADED_POLL_INTERVAL_MS = 5000;

/**
 * FileWatcher monitors a local directory for changes and emits batched
 * FileChangeEvent arrays via the onFlush callback.
 *
 * Features:
 * - Event coalescing: 50ms window + 200ms max-latency cap
 * - Per-path dedup (last event wins within a window)
 * - Default ignore dirs (.git, node_modules, etc.)
 * - .elevoignore support (glob patterns, one per line)
 * - Degraded mode: 5s full_purge polling on watcher errors
 */
export class FileWatcher {
  private watcher: chokidar.FSWatcher | null = null;
  private pending = new Map<string, FileChangeEvent>();
  private flushTimer: ReturnType<typeof setTimeout> | null = null;
  private maxLatencyTimer: ReturnType<typeof setTimeout> | null = null;
  private ignoreRules: string[] = [];
  private closed = false;

  // Degraded mode: when the watcher hits system limits or errors,
  // fall back to sending full_purge every 5 seconds.
  private degraded = false;
  private degradedTimer: ReturnType<typeof setInterval> | null = null;

  constructor(
    private readonly rootDir: string,
    private readonly onFlush: (events: FileChangeEvent[]) => void,
  ) {
    this.ignoreRules = loadElevoIgnore(rootDir);
  }

  /**
   * Start watching. Resolves when the watcher is ready.
   */
  async start(): Promise<void> {
    // Check if directory tree would exceed inotify limits.
    if (shouldDegrade(this.rootDir)) {
      this.enterDegradedMode();
      return;
    }

    const ignoredPaths = buildIgnoredPatterns(this.rootDir);

    this.watcher = chokidar.watch(this.rootDir, {
      ignored: ignoredPaths,
      ignoreInitial: true,
      persistent: true,
      followSymlinks: false,
    });

    this.watcher
      .on('add', (p) => this.handleEvent(p, 'FILE_CHANGE_TYPE_CREATED'))
      .on('addDir', (p) => this.handleEvent(p, 'FILE_CHANGE_TYPE_CREATED'))
      .on('change', (p) => this.handleEvent(p, 'FILE_CHANGE_TYPE_MODIFIED'))
      .on('unlink', (p) => this.handleEvent(p, 'FILE_CHANGE_TYPE_DELETED'))
      .on('unlinkDir', (p) => this.handleEvent(p, 'FILE_CHANGE_TYPE_DELETED'))
      .on('error', (err: unknown) => {
        const error = err instanceof Error ? err : new Error(String(err));
        this.handleWatcherError(error);
      });

    return new Promise<void>((resolve) => {
      this.watcher!.on('ready', () => resolve());
    });
  }

  /**
   * Close the watcher and flush any pending events.
   */
  async close(): Promise<void> {
    this.closed = true;
    this.stopDegradedMode();
    this.flush();
    if (this.watcher) {
      await this.watcher.close();
      this.watcher = null;
    }
  }

  /**
   * Send a full_purge event (empty array signals full purge to caller).
   */
  sendFullPurge(): void {
    this.onFlush([]);
  }

  /** Whether the watcher is in degraded (polling) mode. */
  isDegraded(): boolean {
    return this.degraded;
  }

  private handleEvent(absPath: string, eventType: string): void {
    if (this.closed || this.degraded) return;

    const relPath = path.relative(this.rootDir, absPath);
    if (!relPath || relPath.startsWith('..')) return;

    // Check .elevoignore rules.
    if (this.matchesIgnoreRules(relPath)) return;

    // Per-path dedup: last event wins.
    this.pending.set(relPath, { path: relPath, eventType });

    // Start coalescing timer if not running.
    if (!this.flushTimer) {
      this.flushTimer = setTimeout(() => this.flush(), COALESCE_DELAY_MS);
    }

    // Start max-latency timer if not running.
    if (!this.maxLatencyTimer) {
      this.maxLatencyTimer = setTimeout(() => this.flush(), MAX_LATENCY_MS);
    }
  }

  /**
   * Handle watcher errors. If the error is related to inotify limits,
   * switch to degraded mode.
   */
  private handleWatcherError(err: Error): void {
    if (this.closed) return;

    const msg = err.message || '';
    const isLimitError = msg.includes('no space left on device') ||
                         msg.includes('too many open files') ||
                         msg.includes('ENOSPC');

    if (isLimitError && !this.degraded) {
      console.error(`[FileWatcher] inotify limit reached, switching to degraded mode: ${msg}`);
      this.enterDegradedMode();
    } else {
      console.error(`[FileWatcher] error: ${msg}`);
    }
  }

  /**
   * Enter degraded mode: stop normal watching, start polling with full_purge.
   */
  private enterDegradedMode(): void {
    this.degraded = true;

    // Clear any pending normal events.
    this.pending.clear();
    if (this.flushTimer) { clearTimeout(this.flushTimer); this.flushTimer = null; }
    if (this.maxLatencyTimer) { clearTimeout(this.maxLatencyTimer); this.maxLatencyTimer = null; }

    // Send an immediate full_purge.
    this.sendFullPurge();

    // Start periodic full_purge polling.
    this.degradedTimer = setInterval(() => {
      if (!this.closed) {
        this.sendFullPurge();
      }
    }, DEGRADED_POLL_INTERVAL_MS);

    // Don't block Node.js exit.
    if (typeof this.degradedTimer === 'object' && 'unref' in this.degradedTimer) {
      this.degradedTimer.unref();
    }
  }

  private stopDegradedMode(): void {
    if (this.degradedTimer) {
      clearInterval(this.degradedTimer);
      this.degradedTimer = null;
    }
  }

  private flush(): void {
    if (this.flushTimer) {
      clearTimeout(this.flushTimer);
      this.flushTimer = null;
    }
    if (this.maxLatencyTimer) {
      clearTimeout(this.maxLatencyTimer);
      this.maxLatencyTimer = null;
    }

    if (this.pending.size === 0) return;

    const events = Array.from(this.pending.values());
    this.pending.clear();
    this.onFlush(events);
  }

  private matchesIgnoreRules(relPath: string): boolean {
    const fileName = path.basename(relPath);
    for (const rule of this.ignoreRules) {
      if (matchGlob(rule, fileName) || matchGlob(rule, relPath)) {
        return true;
      }
    }
    return false;
  }
}

/**
 * Build chokidar ignored patterns from default dirs.
 */
function buildIgnoredPatterns(rootDir: string): Array<string | RegExp> {
  const patterns: Array<string | RegExp> = [];
  for (const dir of DEFAULT_IGNORE_DIRS) {
    patterns.push(path.join(rootDir, dir));
    // Also match nested occurrences (e.g. sub/node_modules).
    patterns.push(new RegExp(`(^|[\\/\\\\])${escapeRegExp(dir)}[\\/\\\\]`));
  }
  return patterns;
}

function escapeRegExp(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/**
 * Load .elevoignore rules from the root directory.
 */
function loadElevoIgnore(rootDir: string): string[] {
  const ignorePath = path.join(rootDir, '.elevoignore');
  try {
    const content = fs.readFileSync(ignorePath, 'utf-8');
    return content
      .split('\n')
      .map(line => line.trim())
      .filter(line => line && !line.startsWith('#'));
  } catch {
    return [];
  }
}

/**
 * Simple glob matching (supports * and ? wildcards).
 */
function matchGlob(pattern: string, str: string): boolean {
  const regexStr = '^' + pattern
    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
    .replace(/\*/g, '.*')
    .replace(/\?/g, '.') + '$';
  return new RegExp(regexStr).test(str);
}

/**
 * Check if the directory tree would exceed 80% of the inotify watch limit.
 */
function shouldDegrade(rootDir: string): boolean {
  // Only relevant on Linux.
  try {
    const data = fs.readFileSync('/proc/sys/fs/inotify/max_user_watches', 'utf-8');
    const maxWatches = parseInt(data.trim(), 10);
    if (isNaN(maxWatches) || maxWatches <= 0) return false;

    const dirCount = countDirectories(rootDir);
    return dirCount > (maxWatches * 80 / 100);
  } catch {
    // Not Linux or can't read — proceed optimistically.
    return false;
  }
}

/**
 * Count directories under root, respecting default ignore list.
 */
function countDirectories(root: string): number {
  let count = 0;
  const walk = (dir: string) => {
    try {
      const entries = fs.readdirSync(dir, { withFileTypes: true });
      for (const entry of entries) {
        if (!entry.isDirectory()) continue;
        if (DEFAULT_IGNORE_DIRS.has(entry.name)) continue;
        count++;
        walk(path.join(dir, entry.name));
      }
    } catch {
      // Skip inaccessible directories.
    }
  };
  count++; // Count root itself.
  walk(root);
  return count;
}

File: sdk-go/storage_provider_watch.go
package workspace

import (
	"bufio"
	"log"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/fsnotify/fsnotify"

	pb "github.com/OpenElevo/ElevoWorkspace/sdk-go/proto/workspace/v1"
)

// Default directories to ignore when watching for file changes.
var defaultIgnoreDirs = map[string]bool{
	".git":         true,
	"node_modules": true,
	"__pycache__":  true,
	"target":       true,
	"build":        true,
	".elevo":       true,
}

// fileWatcher monitors the shared directory for file changes and sends
// FileChangedNotification messages through the response channel.
//
// Event coalescing strategy:
//   - First event starts a 50ms timer and records the window start time.
//   - Subsequent events within the window accumulate (per-path dedup, last event wins).
//   - Timer does NOT reset on new events (fixed window, not sliding).
//   - If 200ms have elapsed since window start, flush immediately (max-latency cap).
//   - In degraded mode (inotify limit exceeded), polls every 5s with full_purge=true.
type fileWatcher struct {
	watcher    *fsnotify.Watcher
	rootDir    string
	responseCh chan<- *pb.ClientMessage

	// connDone shares the StorageProvider's per-connection done channel.
	// Closed when the current gRPC connection ends, unblocking trySend
	// so the fileWatcher does not deadlock on a full responseCh.
	connDone *atomic.Value

	// .elevoignore rules.
	ignoreRules []string

	// Event coalescing state.
	pendingEvents map[string]*pb.FileChangeEvent
	mu            sync.Mutex
	timer         *time.Timer
	windowStart   time.Time

	// inotify degraded mode flag.
	degraded atomic.Bool

	// Lifecycle.
	done chan struct{}
}

func newFileWatcher(rootDir string, responseCh chan<- *pb.ClientMessage, connDone *atomic.Value) (*fileWatcher, error) {
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		return nil, err
	}

	fw := &fileWatcher{
		watcher:       watcher,
		rootDir:       rootDir,
		responseCh:    responseCh,
		connDone:      connDone,
		pendingEvents: make(map[string]*pb.FileChangeEvent),
		done:          make(chan struct{}),
	}

	// Load .elevoignore rules if present.
	fw.ignoreRules = loadElevoIgnore(rootDir)

	// Start the event loop immediately.
	go fw.eventLoop()

	// Add watches recursively in the background.
	// Fire-and-forget: scanning a large directory tree can take a long time;
	// the gRPC main loop must start immediately to handle server requests.
	// If the inotify limit is hit during walk, addWatchesRecursive switches to
	// degraded poll mode.
	go func() {
		if err := fw.addWatchesRecursive(rootDir); err != nil {
			log.Printf("[fileWatcher] addWatchesRecursive error: %v", err)
		}
	}()

	return fw, nil
}

// Close stops the file watcher.
func (fw *fileWatcher) Close() {
	close(fw.done)
	fw.watcher.Close()
}

// addWatchesRecursive walks the directory tree and adds fsnotify watches,
// skipping ignored directories.
func (fw *fileWatcher) addWatchesRecursive(root string) error {
	return filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // skip inaccessible paths
		}
		if !info.IsDir() {
			return nil
		}

		name := info.Name()
		if defaultIgnoreDirs[name] && path != root {
			return filepath.SkipDir
		}
		if fw.isIgnored(path) {
			return filepath.SkipDir
		}

		if err := fw.watcher.Add(path); err != nil {
			log.Printf("[fileWatcher] failed to watch %s: %v", path, err)
			// Check if we've hit the inotify limit.
			if isInotifyLimitError(err) {
				fw.degraded.Store(true)
				log.Printf("[fileWatcher] inotify limit reached, switching to degraded mode")
				go fw.degradedPollLoop()
				return filepath.SkipAll
			}
		}
		return nil
	})
}

// eventLoop processes fsnotify events.
func (fw *fileWatcher) eventLoop() {
	for {
		select {
		case <-fw.done:
			return
		case event, ok := <-fw.watcher.Events:
			if !ok {
				return
			}
			fw.handleEvent(event)
		case err, ok := <-fw.watcher.Errors:
			if !ok {
				return
			}
			log.Printf("[fileWatcher] error: %v", err)
		}
	}
}

// handleEvent processes a single fsnotify event, coalescing with pending events.
func (fw *fileWatcher) handleEvent(event fsnotify.Event) {
	if fw.degraded.Load() {
		return // degraded mode handles notifications via polling
	}

	relPath, err := filepath.Rel(fw.rootDir, event.Name)
	if err != nil {
		return
	}

	// Skip ignored paths.
	if fw.isIgnored(event.Name) {
		return
	}

	// Dynamically add watch for new directories.
	if event.Op&fsnotify.Create != 0 {
		if info, err := os.Lstat(event.Name); err == nil && info.IsDir() {
			if !defaultIgnoreDirs[info.Name()] {
				_ = fw.watcher.Add(event.Name)
			}
		}
	}

	changeType := mapFsnotifyOp(event.Op)

	var immediateFlush bool
	fw.mu.Lock()

	fw.pendingEvents[relPath] = &pb.FileChangeEvent{
		Path:      relPath,
		EventType: changeType,
	}

	if fw.timer == nil {
		// First event: start the coalescing window.
		fw.windowStart = time.Now()
		fw.timer = time.AfterFunc(50*time.Millisecond, fw.flush)
	} else if time.Since(fw.windowStart) >= 200*time.Millisecond {
		// Max latency exceeded: mark for immediate flush (outside lock).
		fw.timer.Stop()
		immediateFlush = true
	}
	// else: within window, let existing timer fire.
	fw.mu.Unlock()

	if immediateFlush {
		msg := fw.collectPendingEvents()
		if msg != nil {
			fw.trySend(msg)
		}
	}
}

// flush is called by time.AfterFunc (not under mutex).
func (fw *fileWatcher) flush() {
	msg := fw.collectPendingEvents()
	if msg != nil {
		fw.trySend(msg)
	}
}

// trySend sends a message to responseCh, aborting if the done channel is closed
// or the current gRPC connection has ended. Prevents goroutine leaks when the
// gRPC stream has ended.
func (fw *fileWatcher) trySend(msg *pb.ClientMessage) {
	ch, _ := fw.connDone.Load().(chan struct{})
	select {
	case fw.responseCh <- msg:
	case <-fw.done:
	case <-ch:
	}
}

// collectPendingEvents gathers accumulated events under the lock and returns
// the message to send. Returns nil if there are no pending events.
// Sending is done outside the lock to avoid deadlock when responseCh is full.
func (fw *fileWatcher) collectPendingEvents() *pb.ClientMessage {
	fw.mu.Lock()
	defer fw.mu.Unlock()

	if len(fw.pendingEvents) == 0 {
		fw.timer = nil
		return nil
	}

	events := make([]*pb.FileChangeEvent, 0, len(fw.pendingEvents))
	for _, e := range fw.pendingEvents {
		events = append(events, e)
	}
	fw.pendingEvents = make(map[string]*pb.FileChangeEvent)
	fw.timer = nil

	return &pb.ClientMessage{
		Message: &pb.ClientMessage_FileChanged{
			FileChanged: &pb.FileChangedNotification{Events: events},
		},
	}
}

// degradedPollLoop runs when inotify watches would exceed the limit.
// Sends a full_purge notification every 5 seconds so the server clears all caches.
func (fw *fileWatcher) degradedPollLoop() {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-fw.done:
			return
		case <-ticker.C:
			fw.trySend(&pb.ClientMessage{
				Message: &pb.ClientMessage_FileChanged{
					FileChanged: &pb.FileChangedNotification{
						FullPurge: true,
					},
				},
			})
		}
	}
}

// mapFsnotifyOp converts fsnotify.Op to the proto FileChangeType.
func mapFsnotifyOp(op fsnotify.Op) pb.FileChangeType {
	switch {
	case op&fsnotify.Create != 0:
		return pb.FileChangeType_FILE_CHANGE_TYPE_CREATED
	case op&fsnotify.Remove != 0:
		return pb.FileChangeType_FILE_CHANGE_TYPE_DELETED
	case op&fsnotify.Rename != 0:
		return pb.FileChangeType_FILE_CHANGE_TYPE_RENAMED
	case op&fsnotify.Chmod != 0:
		return pb.FileChangeType_FILE_CHANGE_TYPE_ATTR_CHANGED
	case op&fsnotify.Write != 0:
		return pb.FileChangeType_FILE_CHANGE_TYPE_MODIFIED
	default:
		return pb.FileChangeType_FILE_CHANGE_TYPE_MODIFIED
	}
}

// isIgnored checks if a path should be ignored based on .elevoignore rules
// and default ignore directories.
func (fw *fileWatcher) isIgnored(absPath string) bool {
	relPath, err := filepath.Rel(fw.rootDir, absPath)
	if err != nil {
		return false
	}

	parts := strings.Split(relPath, string(filepath.Separator))
	for _, part := range parts {
		if defaultIgnoreDirs[part] {
			return true
		}
	}

	for _, rule := range fw.ignoreRules {
		matched, _ := filepath.Match(rule, filepath.Base(absPath))
		if matched {
			return true
		}
		// Also try matching the relative path.
		matched, _ = filepath.Match(rule, relPath)
		if matched {
			return true
		}
	}

	return false
}

// loadElevoIgnore reads .elevoignore from the root directory.
// Each line is treated as a filepath.Match pattern.
func loadElevoIgnore(rootDir string) []string {
	f, err := os.Open(filepath.Join(rootDir, ".elevoignore"))
	if err != nil {
		return nil
	}
	defer f.Close()

	var rules []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		rules = append(rules, line)
	}
	return rules
}

// isInotifyLimitError checks if an error is related to inotify watch limits.
func isInotifyLimitError(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return strings.Contains(msg, "no space left on device") ||
		strings.Contains(msg, "too many open files")
}
